dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/jdbc/JdbcProgramOrgUnitAssociationsStore.java

```diff
@@ -64,22 +75,70 @@
 
     private final JdbcTemplate jdbcTemplate;
 
-    public SetValuedMap<String, String> getOrganisationUnitsAssociations( Set<String> uids )
+    private final CacheProvider cacheProvider;
+
+    private final Environment env;
+
+    private Cache<Set<String>> programOrgUnitAssociationCache;
+
+    @PostConstruct
+    public void init()
     {
+        programOrgUnitAssociationCache = cacheProvider.newCacheBuilderForSet( String.class )
+            .forRegion( "pgmOrgUnitAssocCache" )
+            .expireAfterWrite( 1, TimeUnit.HOURS )
+            .withInitialCapacity( 100 )
+            .withMaximumSize( SystemUtils.isTestRun( env.getActiveProfiles() ) ? 0 : 1000 )
+            .build();
+    }
 
-        return jdbcTemplate.query(
-            buildSqlQueryForRawAssociation( uids ),
-            resultSet -> {
-                SetValuedMap<String, String> setValuedMap = new HashSetValuedHashMap<String, String>();
-                while ( resultSet.next() )
-                {
-                    setValuedMap.putAll(
-                        resultSet.getString( 1 ),
-                        Arrays.asList( (String[]) resultSet.getArray( 2 ).getArray() ) );
-
-                }
-                return setValuedMap;
-            } );
+    @EventListener
+    public void handleApplicationCachesCleared( ApplicationCacheClearedEvent event )
+    {
+        programOrgUnitAssociationCache.invalidateAll();
+    }
+
+    public SetValuedMap<String, String> getOrganisationUnitsAssociations( Set<String> uids )
+    {
+        SetValuedMap<String, String> setValuedMap = new HashSetValuedHashMap<String, String>();
+        boolean cached = true;
+        for ( String uid : uids )
+        {
+            Optional<Set<String>> orgUnitUids = programOrgUnitAssociationCache.get( uid );
+            if ( !orgUnitUids.isPresent() )
```

Comment:
@ameenhere just wondering, do we store an empty Set in those cases where a program has zero org unit associations? That might be for good reasons if the org unit does not offer any services at this time. If not, lets assume 20 programs are requested, where 19 programs have thousands of org units associated, and 1 program has no org units. Will we _always_ bypass the cache and go to the database in this case?